realm/realm-library/src/objectServer/java/io/realm/RealmUser.java

```diff
@@ -218,42 +225,211 @@ public State getState() {
-     * <p>
-     * If the user is already logged out, this method does nothing.
-     * <p>
-     * This method should be called whenever the application is committed to not using a user again.
-     * Failing to call this method may result in unused files and metadata needlessly taking up space.
+     * Log the user out of the Realm App asynchronously. This will unregister them on the device, stop any
+     * synchronization to and from the users Realms, and those Realms will be deleted next time
+     * the app restarts. Therefor logging out should not be done until all changes to Realms have
+     * been uploaded to the server.
      * <p>
      * Once the Realm App has confirmed the logout any registered {@link AuthenticationListener}
      * will be notified and user credentials will be deleted from this device.
+     * <p>
+     * Logging out anonymous users will remove them immediately instead of marking them as
+     * {@link RealmUser.State#LOGGED_OUT}. All other users will be marked as {@link RealmUser.State#LOGGED_OUT}
+     * and will still be returned by {@link #allUsers()}. They can be removed completely by calling
+     * {@link #removeUser()}.
      *
-     * @throws IllegalStateException if not called on a looper thread.
+     * @param callback callback when logging out has completed or failed. The callback will always
+     * happen on the same thread as this method is called on.
+     * @throws IllegalStateException if called from a non-looper thread.
+     */
+    public RealmAsyncTask logOutAsync(RealmApp.Callback<RealmUser> callback) {
+        final RealmUser user = this;
+        Util.checkLooperThread("Asynchronous log out is only possible from looper threads.");
+        return new RealmApp.Request<RealmUser>(SyncManager.NETWORK_POOL_EXECUTOR, callback) {
+            @Override
+            public RealmUser run() throws ObjectServerError {
+                logOut();
+                return user;
+            }
+        }.start();
+    }
+
+    /**
+     * Returns a wrapper for managing API keys controlled by the current user.
+     *
+     * @return wrapper for managing API keys controlled by the current user.
+     * @throws IllegalStateException if no user is currently logged in.
+     */
+    public synchronized ApiKeyAuthProvider getApiKeyAuthProvider() {
+        checkLoggedIn();
+        if (apiKeyAuthProvider == null) {
+            apiKeyAuthProvider = new ApiKeyAuthProvider(this);
+        }
+        return apiKeyAuthProvider;
+    }
+
+    /**
+     * FIXME Add support for functions. Name of Class and method still TBD.
+     */
+    public RealmFunctions getFunctions() {
+        return null;
+    }
+
+    /**
+     * FIXME Add support for push notifications. Name of Class and method still TBD.
      */
-    public RealmAsyncTask logOutAsync(RealmApp.Callback callback) {
-        return app.logOutAsync(this, callback);
+    public RealmPushNotifications getFSMPushNotifications() {
```

Comment:
I thought we renamed this in another PR to lose the prefix?

Reply:
I think you are right. I'll rename here.